api/src/main/java/io/opentelemetry/common/Attributes.java

```diff
@@ -0,0 +1,251 @@
+   * Duplicate keys will be removed.
+   */
+  public static Attributes of(
+      String key1,
+      AttributeValue value1,
+      String key2,
+      AttributeValue value2,
+      String key3,
+      AttributeValue value3) {
+    return sortAndFilter(key1, value1, key2, value2, key3, value3);
+  }
+
+  /**
+   * An {@link Attributes} instance with four key-value pairs. Order of the keys is not preserved.
+   * Duplicate keys will be removed.
+   */
+  public static Attributes of(
+      String key1,
+      AttributeValue value1,
+      String key2,
+      AttributeValue value2,
+      String key3,
+      AttributeValue value3,
+      String key4,
+      AttributeValue value4) {
+    return sortAndFilter(key1, value1, key2, value2, key3, value3, key4, value4);
+  }
+
+  /**
+   * An {@link Attributes} instance with five key-value pairs. Order of the keys is not preserved.
+   * Duplicate keys will be removed.
+   */
+  public static Attributes of(
+      String key1,
+      AttributeValue value1,
+      String key2,
+      AttributeValue value2,
+      String key3,
+      AttributeValue value3,
+      String key4,
+      AttributeValue value4,
+      String key5,
+      AttributeValue value5) {
+    return sortAndFilter(
+        key1, value1,
+        key2, value2,
+        key3, value3,
+        key4, value4,
+        key5, value5);
+  }
+
+  /** Creates a new {@link Builder} instance for creating arbitrary {@link Attributes}. */
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  @AutoValue
+  @Immutable
+  abstract static class ArrayBackedAttributes extends Attributes {
+    abstract List<Object> data();
+
+    ArrayBackedAttributes() {}
+
+    @Override
+    public void forEach(KeyValueConsumer<AttributeValue> consumer) {
+      for (int i = 0; i < data().size(); i++) {
+        consumer.consume((String) data().get(i), (AttributeValue) data().get(++i));
+      }
+    }
+  }
+
+  /**
+   * Enables the creation of an {@link Attributes} instance with an arbitrary number of key-value
+   * pairs.
+   */
+  public static class Builder {
+    private final List<Object> data = new ArrayList<>();
+
+    /** Create the {@link Attributes} from this. */
+    public Attributes build() {
+      return sortAndFilter(data.toArray());
+    }
+
+    /**
+     * Add a bare {@link AttributeValue} to this.
+     *
+     * @return this Builder
+     */
+    public Builder addAttribute(String key, AttributeValue value) {
+      data.add(key);
+      data.add(value);
+      return this;
+    }
+
+    /**
+     * Add a String {@link AttributeValue} to this.
+     *
+     * @return this Builder
+     */
+    public Builder addAttribute(String key, String value) {
+      data.add(key);
+      data.add(stringAttributeValue(value));
+      return this;
+    }
+
+    /**
+     * Add a long {@link AttributeValue} to this.
+     *
+     * @return this Builder
+     */
+    public Builder addAttribute(String key, long value) {
```

Comment:
Thanks! The API in this PR LGTM. Only comment is may stick with `putAttribute` instead of `add`, more familiar since it's inspired by Java's `Map` and conveys that duplicates are deduped.

Reply:
I'm wondering if we should have `setAttribute()` instead, to align this with `Span.setAttribute()`. No strong feeling though.

Reply:
I'm totally fine, either way. @anuraaga , what do you think about `setAttribute`, rather than `putAttribute` ?

Reply:
+1 for `setAttribute`, to align with `Span.setAttribute`

Reply:
ok, I'll assume @anuraaga is ok, and if not, we can change it.

Reply:
done.

Reply:
Sorry for the late SGTM